src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs

```diff
@@ -238,6 +238,24 @@ public static bool AreSame<T>([AllowNull] ref T left, [AllowNull] ref T right)
             // ret
         }
 
+        /// <summary>
+        /// Reinterprets the given reference as a reference to a value of type <typeparamref name="TTo"/>.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The size of <typeparamref name="TFrom" /> and <typeparamref name="TTo" /> are not the same.</exception>
+        [Intrinsic]
+        [NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static TTo BitCast<TFrom, TTo>(TFrom source)
+            where TFrom : struct
+            where TTo : struct
+        {
```

Comment:
Should this check `IsReferenceOrContainsReferences` on `TFrom`/`TTo` and throw?

Reply:
Simply checking that `IsReferenceOrContainsReferences` is the same between `TFrom` and `TTo` will only catch a subset of cases. It won't catch the more complex cases where they match, but the layouts are then considered "incompatible".

The implementation will "do the right thing" as the software fallback just uses `As<TFrom, TTo>(ref value)` and the intrinsic logic checks for compatible layouts. It's just that them being actually incompatible may result in UB, much as it would for `As` or many of the other `Unsafe` APIs.